golem-cli/src/model/application_manifest.rs

```diff
@@ -0,0 +1,919 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use golem_common::model::{ComponentType, InitialComponentFilePath, InitialComponentFilePathAndPermissions, InitialComponentFilePermissions};
+use golem_wasm_rpc_stubgen::model::oam;
+use golem_wasm_rpc_stubgen::model::unknown_properties::{HasUnknownProperties, UnknownProperties};
+use golem_wasm_rpc_stubgen::model::validation::{ValidatedResult, ValidationBuilder};
+use golem_wasm_rpc_stubgen::model::wasm_rpc::{
+    include_glob_patter_from_yaml_file, DEFAULT_CONFIG_FILE_NAME, OAM_COMPONENT_TYPE_WASM, OAM_COMPONENT_TYPE_WASM_BUILD, OAM_COMPONENT_TYPE_WASM_RPC_STUB_BUILD, OAM_TRAIT_TYPE_WASM_RPC,
+};
+use glob::glob;
+use golem_wasm_rpc_stubgen::commands::declarative::ApplicationResolveMode;
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use url::Url;
+use std::collections::{BTreeMap, BTreeSet};
+use std::path::{Path, PathBuf};
+use golem_wasm_rpc_stubgen::model::oam::TypedTraitProperties;
+use tracing::warn;
+
+use super::GolemError;
+
+pub fn load_app(app_resolve_mode: &ApplicationResolveMode) -> Result<ApplicationManifest, GolemError> {
```

Comment:
This is mostly taken from wasm-rpc-stubgen and extended. I think it would be good to merge and move them into some common crate so the validation / import / etc. logic is consistent across projects.

Reply:
wasm-rpc-stubgen is a dependency even currently (the rpc cli commands are even "embedded" into the other golem-cli commands), and most of the things should be public for reuse. The intention is to have the generic oam and rpc manifest model available and working in wasm-rpc, and only call and extend it (for the extra properties and validation) here. It might be happen that we have to make some thing more generic of public in wasm-rpc, but currently the plan is to reuse from that lib directly.

Reply:
The problem is that a lot of the stuff used here is private. The main offenders are

* [load-app](https://github.com/golemcloud/wasm-rpc/blob/06e127aeb10bc602a7920c53b5c03f6accd54421/wasm-rpc-stubgen/src/commands/declarative.rs#L322C4-L322C19)
* [the manifest itself](https://github.com/golemcloud/wasm-rpc/blob/06e127aeb10bc602a7920c53b5c03f6accd54421/wasm-rpc-stubgen/src/model/wasm_rpc.rs#L57) -- This guy has all the validation logic and the functionality to import referenced wasmbuild projects. Sadly the underlying [oam manifest with extra properties](https://github.com/golemcloud/wasm-rpc/blob/06e127aeb10bc602a7920c53b5c03f6accd54421/wasm-rpc-stubgen/src/model/oam.rs#L150) is not exposed from there. We should absolutely refactor this part once those are public though.

Reply:
We have a problem of code duplication already in some other places, and we are working on getting rid of them, so I would rather have this fixed first in wasm-rpc and applied here. Even if it's another repo, it is part of the project.

Reply:
I understand your reasoning here 👍.
Though if possible, I would like to do this as a followup, given that it doesn't change the behaviour and the pr is already huge (what I mostly am a bit afraid of is other things getting merged and having to rebase this monstrosity). Doing a dance with releases of multiple repos would further complicate it.

Reply:
In that case can you please add a few TODOs for it (to any source that was copied)?
As for a followup, I think you can let that go, as there will be soon a bigger batch of wasm-rpc-stubgen updates, so it can be part of that.